rotkehlchen/balances/historical.py

```diff
@@ -115,23 +117,31 @@ def get_erc721_tokens_balances(
         return {asset.resolve_to_evm_token(): balance for asset, balance in current_balances.items() if balance > ZERO}  # noqa: E501
 
     def get_asset_balance(self, asset: Asset, timestamp: Timestamp) -> HistoricalBalance:
-        """Get historical balance for a single asset at a given timestamp
+        """Get historical balance for a single asset at a given timestamp.
+
+        The inner query gets the latest balance per bucket via MAX(timestamp + sequence_index),
+        relying on SQLite's bare column behavior to return non-aggregated columns from that row.
+        See https://www.sqlite.org/lang_select.html#bareagg
 
         May raise:
         - NotFoundError if balance for the asset at the given timestamp does not exist.
-        - DeserializationError if there is a problem deserializing an event from DB.
         """
-        events, main_currency = self._get_events_and_currency(to_ts=timestamp, assets=(asset,))
-        if len(events) == 0:
-            raise NotFoundError(f'No historical data found for {asset} until the given timestamp.')
+        with self.db.conn.read_ctx() as cursor:
+            main_currency = self.db.get_setting(cursor=cursor, name='main_currency')
+            cursor.execute(
+                """SELECT SUM(metric_value) FROM (SELECT em.metric_value, MAX(he.timestamp + he.sequence_index)
+                    FROM event_metrics em
+                    INNER JOIN history_events he ON em.event_identifier = he.identifier
+                    WHERE em.metric_key = ? AND he.timestamp <= ? AND he.asset = ?
+                    GROUP BY he.location, he.location_label, em.protocol
+                )""",  # noqa: E501
+                (EventMetricKey.BALANCE.serialize(), ts_sec_to_ms(timestamp), asset.identifier),
+            )
+            result = cursor.fetchone()
```

Comment:
```suggestion
            ).fetchone()[0]
```
no need for this to be a seperate line. Also can do the [0] here instead of in the line below I believe.